### ISSUE TITLE
Qualified type of constructor invocations

### ISSUE THREAD

**Issue by mernst** [typetools/checker-framework#2163]
The qualified type of the constructor invocation `new MyClass(22)` depends on three annotations:

```
class @A MyClass {
  @B MyClass(int arg) { ... }
}

class Client {
  ... new @C MyClass(22) ...
}
```

These three annotations must have the following relationship:
```
  @C <: @B <: @A
```
If any of these relationships is violated, that is an error.

The qualified type of the constructor invocation is `@C`.  (That is also equal to the greatest lower bound of `@A`, `@B`, and `@C`.)

If `@C` is omitted, then the type of `new MyClass(22)` is `@B MyClass`.
If `@C` and `@B` are omitted, then the type of `new MyClass(22)` is `@A MyClass`, just like any other use of `MyClass` without an explicit annotation.
If `@B` is omitted, then the result type of the constructor ` MyClass(int arg) { ... }` and the type  of `this` inside the constructor are `@A MyClass`.  This is the same for any other use of `MyClass` without an explicit annotation.

Currently, it seems that the Checker Framework may sometimes silently ignore the annotation `@C` (see https://github.com/typetools/checker-framework/issues/2159).
The Checker Framework should never ignore any annotation.  If an annotation is inconsistent with other annotations, the Checker Framework should issue an error.

**Comment by smillst** [comment]
>The qualified type of the constructor invocation is @C. (That is also equal to the least upper bound of @A, @B, and @C.)

You mean GLB?

>If an annotation is omitted from source code, it should default to a value consistent with the other annotations.

This paragraph should be clarified:

If `@C` is omitted, then the type of `new MyClass(...)` is `@B MyClass`.
If `@C` and `@B` are omitted, then the type of `new MyClass(...)` is `@A MyClass`, just like any other use of `MyClass` without an explicit annotation.
If `@B` is omitted, then the resulting type of the constructor ` MyClass(int arg) { ... }` and the type  of `this` inside the constructor are `@A MyClass`.  This is the same for any other use of `MyClass` without an explicit annotation.

**Comment by t-rasmud** [comment]
@mernst I have added test cases for this issue [here](https://github.com/t-rasmud/checker-framework/blob/issue32/checker/tests/determinism/Issue2163.java).

**Comment by t-rasmud** [comment]
The correct relationship between the the annotations `@A`, `@B`, and `@C` is as follows:
`@B <: @A` and `@B <: @C`
The checker framework checks for the relationship `@B <: @A` and issues an error if it is violated.
The relationship `@B <: @C` gets violated when the explicit type at constructor invocation is a subtype of the constructor declaration type. Example: `new @C MyClass(22)` where `@C` is a subtype of `@B`. Currently, this does not result in an error. This should be treated like a downcast [`(@C MyClass) new MyClass(22)`] and the checker framework should issue a warning (not an error).
